create a pg to pgml convert script

pstaabp opened the pull request:
This is a script to convert a PG problem to PGML format.  Details are in the script file.  Overall, to run this, it is
```
convert-to-pgml.pl file1.pg file2.pg
```
converts each of the files to PGML format and creates a `.bak` file.

pstaabp commented:
This is ready for testing.  I've made the script more robust and improved documentation.  If there are other features/conversions requested, happy to take those.

drgrice1 commented on lib/WeBWorK/PG/ConvertToPGML.pm:
```diff
@@ -0,0 +1,182 @@
+
+our @EXPORT = qw(convertToPGML);
+
+use Data::Dumper;
```
Please delete this and the `Dumper` call on line 81.

drgrice1 commented on lib/WeBWorK/PG/ConvertToPGML.pm:
```diff
@@ -0,0 +1,182 @@
+Converts a pg file to PGML format.
+
+=head1 OPTIONS
+
+
```
Please delete the extra empty line here.

drgrice1 reviewed (commented):
This is a start, but there seem to be several things that aren't working very well.

I started with Library/UMN/calculusStewartCCC/s_2_3_12.pg

and got (I also manually deleted the comment lines):
```
DOCUMENT();
loadMacros('PGstandard.pl', 'PGML.pl', , 'PGcourse.pl');

Context("Numeric");

$a = list_random(-3,-2,-1,1,5);

$ans = Compute("(2*($a)^2+(4*$a)-5)/(($a)^2-(5*$a)+6)")->reduce;

BEGIN_PGML
Evaluate the limit, if it exists.  If a limit [$BITALIC] does not exist, [$EITALIC] type "DNE".
[``
\lim_{x \to [$a]} \frac{2x^2+4x-5}{x^2-5x+6}
``]

---

Limit: [_]{}{20}  
END_PGML

# ANS( $ans->cmp );

ENDDOCUMENT();
```
Observe the following:
* There are now two consecutive commas in the `loadMacros` call.
* The `$BITALIC` and `$EITALIC` did not get converted.
* There is whitespace added at the end of the `Limit: [_]{}{20}` line.
* The ```[``\lim ... ``]``` was initially in display style math mode with `\[ ... \]`, but was converted to inline display style math mode.  To get the same math mode it should be converted to use the three backtick version ````[```\lim ... ```]````.

Note that if you run the scrip on this file again, the two consecutive commas in the `loadMacros` call are turned into `, '',`.  That is the empty string is added in between those two commas.  This causes an error if the problem file is used (assuming you fix the answer).

Also, could you make the comments in the added files consistent.  Some comments are capitalized and properly punctuated, but others are not.

pstaabp commented:
> There is whitespace added at the end of the Limit: [_]{}{20} line.

This is added because there is a `$BR` at the end of the line and using the two spaces for a line break in PGML. 

I think original `$BR` shouldn't be there.

Fixed these errors and added the removal of any lines with only comments on them.

drgrice1 reviewed (commented):
Starting to look closely at this.  There is a lot of clean up needed in this.

pstaabp commented on lib/WeBWorK/PG/ConvertToPGML.pm:
```diff
@@ -0,0 +1,181 @@
+				$macros .= $mrow[0] if $mrow[0] !~ /^\s*$/;
+			}
+			# Split by commas and pull out the quotes.
+			my @macros = map {s/['"\s]//gr} split(/\s*,\s*/, $macros =~ s/loadMacros\((.*)\)\;$/$1/r);
+			@macros = grep {
+				$_ !~ /(PGstandard|PGML|PGauxiliaryFunctions|PGbasicmacros|PGanswermacros|MathObjects|PGcourse).pl/
+			} @macros;
+			@macros = grep { $_ !~ /^#/ } @macros;
```
I was thinking of doing this.

drgrice1 reviewed (approved):
I think this will need some more tweaking, but this is a start.

drgrice1 commented on lib/WeBWorK/PG/ConvertToPGML.pm:
```diff
@@ -0,0 +1,210 @@
+				map {s/['"\s]//gr}
+				split(/\s*,\s*/, $macros =~ s/loadMacros\((.*)\)\;$/$1/r);
+
+			# my @macros = map {s/['"\s]//gr} split(/\s*,\s*/, $macros =~ s/loadMacros\((.*)\)\;$/$1/r);
+			# @macros = grep {
+			# 	$_ !~ /(PGstandard|PGML|PGauxiliaryFunctions|PGbasicmacros|PGanswermacros|MathObjects|PGcourse).pl/
+			# } @macros;
+			# @macros = grep { $_ !~ /^#/ } @macros;
```
Delete this commented out code.  It seems to be a remnant of the previous code from the preceding lines.

pstaabp replied:
done

drgrice1 reviewed (commented):
Either the things commented on need to be fixed, or the help comments shown in the webwork2 problem editor for this need to be made much more far reaching.  Clearly not only answer rules are not dealt with correctly.

pstaabp commented on lib/WeBWorK/PG/ConvertToPGML.pm:
```diff
@@ -0,0 +1,210 @@
+
+sub convertANSrule {
+	my ($str) = @_;
+	if ($str =~ /(.*)\\\{\s*((\$\w+)->)?ans_rule(\((\d+)\))?\s*\\\}(.*)$/) {
```
Updated the comment.    I didn't check for things like `named_ans_rule`, which are less common.  Also, I think trying to convert arrays and hashes will be difficult.
